### DISCUSSION TITLE
[v2] Should all transforms work with ParameterRange?

### DISCUSSION THREAD

**Discussion by eigenvivek** [TorchIO-project/torchio#1472] [Q&A] [answered]
I've been trying out the v2 branch and really enjoying the changes (especially the ability to move subjects to the GPU)! I'm confused if every transform is supposed to work with the `ParameterRange`, or if some are still going to be purely deterministic.

In particular, I was trying to use the `ParameterRange` in conjunction with `Resample`. I could get random resampling if I do something like this
```
# Works
target = ParameterRange(value=(2, 4, 2, 4, 3, 6)).sample()
_ = Resample(target=target)(image)
```

but not this
```
# Doesn't work
_ = Resample(target=ParameterRange(value=(2, 4, 2, 4, 3, 6)))(image)
```

If there are some augs that aren't supposed to work with ParameterRange it would be very helpful to document that. Reading the [docs](https://smokeshow.helpmanual.io/025f2b6c3x1z5m1p6d20/tutorials/augmentation/#summary) gave me the impression that they all did.

Thanks so much for all your work on the new version, it has a lot of features/improvements that are really fun to use!

**Comment by romainVala** [comment]
nice to ear about the v2, it does look promising !!
I hope to have time soon to test it 
big thanks to Fernando !

**Comment by fepegar** [comment]
Hi both. Thanks so much for the feedback and your kind words.

`v2` is extremely experimental and it's still early so now is the time to try, go crazy, give feedback and request features.

@eigenvivek I think what you're proposing should work (I mean I want it to work in the release so will try to fix it).

**Comment by fepegar** [comment, marked as answer]
@eigenvivek I've made some changes. Summarised by the clanker below:

---

Short version: in v2 you pass the **spec directly**; you don't construct a
`ParameterRange` yourself. So this is the intended style:

```python
Resample(target=(2, 4, 2, 4, 3, 6))(image)   # per-axis spacing ranges
Resample(target=(1, 2))(image)               # uniform spacing range
Affine(degrees=(-10, 10))(image)             # same convention everywhere
Affine(degrees=Choice([-90, 0, 90]))(image)  # discrete
```

A couple of clarifications prompted by your report:

1. **`ParameterRange` has become an internal class (`_ParameterRange`).** It's the wrapper
   transforms build for you from the raw spec, and passing an instance never
   actually worked (e.g. `Affine(degrees=ParameterRange((-10, 10)))` errored
   too). So rather than make it accept itself, we're removing it from the
   public API. The public building blocks are: a scalar (fixed), a 2-tuple
   `(a, b)` (uniform), a 3-/6-tuple (per-axis values/ranges), `Choice` for a
   discrete set, or any `torch.distributions.Distribution`. The `.sample()`
   workaround you found is no longer needed.

2. **`Resample(target=...)` now supports random spacing**, which is what you
   were really after. A scalar/3-tuple stays deterministic; a 2-tuple,
   6-tuple, `Choice`, or `Distribution` is sampled once per call at apply
   time. (Note: a 2-tuple like `(2, 4)` samples each axis independently, so
   the result is generally anisotropic.)

3. **Not every parameter is randomizable, and that's by design.**
   Value/range/distribution params (e.g. `degrees`, `scales`, `translation`,
   `max_displacement`, and now `Resample`'s spacing `target`) are; *structural*
   ones (interpolation, `center`, or `target` given as an image/path/
   shape-affine) are not. We're tightening the docs so this is explicit.

This is on the `v2` branch now. Thanks again for kicking the tires.

---

Back to the human. This should work on `v2` (which will soon be `main`).

```python
import torch
from torch.distributions import Uniform

import torchio as tio
from torchio import Choice

torch.manual_seed(0)
colin = tio.datasets.Colin27(version=1998)
image = colin.t1
print("shape:  ", image.shape)
print("spacing:", image.spacing)
image.plot()
```

```
shape:   (1, 181, 217, 181)
spacing: (1.0, 1.0, 1.0)
```

<img width="1269" height="464" alt="image" src="https://github.com/user-attachments/assets/7a988dd6-48b5-4d5a-aba9-8fecabab08c6" />

```python
print("scalar :", tio.Resample(2)(image).spacing)
print("3-tuple:", tio.Resample(target=(1, 2, 3))(image).spacing)
```

```
scalar : (2.0, 2.0, 2.0)
3-tuple: (1.0, 2.0, 3.0)
```

```python
# Per-axis ranges: (lo1, hi1, lo2, hi2, lo3, hi3)
for _ in range(3):
    resampled = tio.Resample(target=(2, 4, 2, 4, 3, 6))(image)
    spacing = tuple(round(s, 3) for s in resampled.spacing)
    print(f"spacing={spacing}  ->  shape={resampled.shape}")
```

```
spacing=(2.264, 2.615, 4.902)  ->  shape=(1, 79, 82, 36)
spacing=(3.793, 2.911, 4.897)  ->  shape=(1, 47, 74, 36)
spacing=(2.803, 2.045, 3.507)  ->  shape=(1, 64, 106, 51)
```

```python
# Uniform range (same range every axis, sampled independently)
print("range  :", tuple(round(s, 3) for s in tio.Resample(target=(1, 2))(image).spacing))
# Discrete set
print("Choice :", tio.Resample(target=Choice([1.0, 2.0]))(image).spacing)
# Any distribution
print("Uniform:", tuple(round(s, 3) for s in tio.Resample(target=Uniform(1.0, 2.0))(image).spacing))
```

```
range  : (1.519, 1.698, 1.8)
Choice : (1.0, 2.0, 1.0)
Uniform: (1.176, 1.27, 1.151)
```

Is this helpful? Do you have any suggestions for improvement?

**Reply by eigenvivek** [reply]
Just fetched the new main branch and tried this out - works exactly as expected! Thank you very much.

**Reply by fepegar** [reply]
Excellent! Thanks for trying.
